### ISSUE TITLE
Q4KExperts GGUF produces NaNs from layer 0 on Metal; Q2 works

### ISSUE THREAD

**Issue by ensemble-learning** [antirez/ds4#65]
### Summary

I am testing `ds4` with the DeepSeek V4 Flash GGUF files from:

- Repo: `antirez/deepseek-v4-gguf`
- Q2 file: `DeepSeek-V4-Flash-IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8-chat-v2.gguf`
- Q4 file: `DeepSeek-V4-Flash-Q4KExperts-F16HC-F16Compressor-F16Indexer-Q8Attn-Q8Shared-Q8Out-chat-v2.gguf`

The Q2 model works correctly, but the Q4 model loads and runs while producing invalid activations/logits. Generation degenerates to repeated BOS tokens:

```text
<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>...
```

With `--dump-logprobs`, Q4 has no finite top logprobs, while Q2 has normal finite logits.

### Important note: downloaded file size appears correct

I initially suspected a corrupted or incomplete Q4 download, but the file size looks correct when accounting for GB vs GiB units.

The local Q4 file size was:

```
165102258845 bytes
```

That is:

```
165.10 GB decimal
153.76 GiB binary
```

Hugging Face / browser UIs usually display decimal GB, while tools like `du -h`, `ls -lh`, or the program output may show binary GiB. So a file shown as about `153.76 GiB` locally corresponds to about `165.10 GB`, which matches the expected Hugging Face size. Therefore the size alone does not indicate an incomplete download.

### Environment

- Machine: Apple Silicon Mac Studio M3 Ultra
- RAM: 512 GB
- OS/backend: macOS + Metal
- Build command:

```
make ds4
```

### Q2 works

Using the Q2 GGUF:

```
./ds4 \

  -m gguf/DeepSeek-V4-Flash-IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8-chat-v2.gguf \

  --ctx 512 \

  -p 'Hello, write one short sentence.' \

  --nothink \

  --temp 0 \

  -n 16
```

This produces normal text, for example:

```
The sun dipped below the horizon, painting the sky in shades of amber and rose
```

Q2 finite diagnostics for layer 0 also look normal:

```
finite hc_attn_pre layer 0 ... finite=40960 nan=0
finite attn_norm   layer 0 ... finite=40960 nan=0
finite kqv_out     layer 0 ... finite=327680 nan=0
finite attn_out    layer 0 ... finite=40960 nan=0
finite ffn_out     layer 0 ... finite=40960 nan=0
```

### Q4 fails

Using the Q4 GGUF:

```
./ds4 \

  -m gguf-bak/DeepSeek-V4-Flash-Q4KExperts-F16HC-F16Compressor-F16Indexer-Q8Attn-Q8Shared-Q8Out-chat-v2.gguf \

  --ctx 512 \

  -p 'Hello' \

  --nothink \

  --temp 0 \

  -n 1
```

Output:

```
<｜begin▁of▁sentence｜>
```

Longer generation repeats BOS.

### Inspect output

Q4 inspect succeeds and identifies the tensor types:

```
model: DeepSeek V4 Flash
arch:  deepseek4
gguf:  v3, 58 metadata keys, 1328 tensors
layers: 43
attention: heads=64 kv_heads=1 head_dim=512 swa=128
indexer: heads=64 head_dim=128 top_k=512
experts: count=256 used=6 groups=0 groups_used=0
file size: 153.76 GiB
tensor bytes described by GGUF: 153.32 GiB
logical parameters: 284.33 B
tensor types:
  f32        492 tensors, 0.00 GiB
  f16        359 tensors, 2.04 GiB
  q8_0       345 tensors, 6.15 GiB
  q4_k       129 tensors, 145.12 GiB
  i32          3 tensors, 0.01 GiB
```

Q2 inspect:

```
tensor types:
  f32        492 tensors, 0.00 GiB
  f16        359 tensors, 2.04 GiB
  q8_0       345 tensors, 6.15 GiB
  q2_k        43 tensors, 28.22 GiB
  iq2_xxs     86 tensors, 44.34 GiB
  i32          3 tensors, 0.01 GiB
```

### Added finite/NaN diagnostics

I added temporary diagnostics guarded by an env var:

```
DS4_METAL_GRAPH_FINITE_TRACE=1

DS4_METAL_GRAPH_FINITE_LAYER=0
```

Then ran Q4:

```
DS4_METAL_GRAPH_FINITE_TRACE=1 \

DS4_METAL_GRAPH_FINITE_LAYER=0 \

./ds4 \

  -m gguf-bak/DeepSeek-V4-Flash-Q4KExperts-F16HC-F16Compressor-F16Indexer-Q8Attn-Q8Shared-Q8Out-chat-v2.gguf \

  --ctx 512 \

  -p 'Hello' \

  --nothink \

  --temp 0 \

  -n 1
```

Q4 layer 0 diagnostic result:

```
finite hc_attn_flat layer 0 pos 0 n=163840 finite=49152 nan=114688
finite hc_attn_mix  layer 0 pos 0 n=240    finite=72    nan=168
finite hc_attn_pre  layer 0 pos 0 n=40960  finite=12288 nan=28672
finite attn_norm    layer 0 pos 0 n=40960  finite=12288 nan=28672
finite kqv_out      layer 0 pos 0 n=327680 finite=0     nan=327680
finite attn_out     layer 0 pos 0 n=40960  finite=0     nan=40960
finite hc_attn_post layer 0 pos 0 n=163840 finite=0     nan=163840
finite hc_ffn_pre   layer 0 pos 0 n=40960  finite=0     nan=40960
```

This shows the failure occurs extremely early: before FFN/MoE. In fact, NaNs already appear in `hc_attn_flat`, which is derived from the initial HC state before layer 0 attention.

I also tried forcing the GPU batch embedding path:

```
DS4_METAL_GPU_BATCH_EMBED_MIN=1
```

The Q4 diagnostic output was unchanged.

### Current hypothesis

The first observed failure is not in the Q4_K MoE expert kernel. The NaNs appear before MoE is reached.

The earliest failing component seems to be one of:

1. `token_embd.weight` data for the Q4 GGUF,
2. GGUF tensor offset/layout interpretation for the Q4 file,
3. F16 embedding read/expansion path,
4. Metal mapped model view / offset calculation for this particular Q4 GGUF.

Because Q2 uses the same code path and works, this appears specific to the Q4 GGUF layout/data.

### Expected behavior

Q4 should generate finite logits and normal text, similar to Q2.

### Actual behavior

Q4 produces NaNs from layer 0 and generation collapses to BOS token repetition.

### Repro commands

```
make ds4



./ds4 --inspect \

  -m gguf-bak/DeepSeek-V4-Flash-Q4KExperts-F16HC-F16Compressor-F16Indexer-Q8Attn-Q8Shared-Q8Out-chat-v2.gguf



DS4_METAL_GRAPH_FINITE_TRACE=1 \

DS4_METAL_GRAPH_FINITE_LAYER=0 \

./ds4 \

  -m gguf-bak/DeepSeek-V4-Flash-Q4KExperts-F16HC-F16Compressor-F16Indexer-Q8Attn-Q8Shared-Q8Out-chat-v2.gguf \

  --ctx 512 \

  -p 'Hello' \

  --nothink \

  --temp 0 \

  -n 1
```

### Additional note

I am currently re-downloading the Q4 GGUF to fully rule out silent corruption, but based on file size the previous download did not look truncated: `165102258845 bytes = 165.10 GB = 153.76 GiB`.

**Comment by jelspace** [comment]
Check file correctness with sha256sum , type in terminal:  sha256sum filename

sha256 checksums from hf website are:

DeepSeek-V4-Flash-IQ2XXS-w2Q2K-AProjQ8-SExpQ8-OutQ8-chat-v2.gguf
SHA256 is:  31598c67c8b8744d3bcebcd19aa62253c6dc43cef3b8adf9f593656c9e86fd8c

DeepSeek-V4-Flash-Q4KExperts-F16HC-F16Compressor-F16Indexer-Q8Attn-Q8Shared-Q8Out-chat-v2.gguf
SHA256 is:  39e5de72ac544fdd5ffaf83ec28e36aaf3341b145235488e67d59400bbb3af55

**Comment by ensemble-learning** [comment]
Thanks, apparently, my q2 weight is right. but q4 weight is wrong. I will re-download.
